src/main/scala/com/fulcrumgenomics/vcf/filtration/FilterSomaticVcf.scala

```diff
@@ -24,190 +24,182 @@
+      filters = source.header.filters ++ SomaticFilters.flatMap(_.VcfFilterLines),
+    )
 
-    val builder = new PileupBuilder(bamIn.dict, mappedPairsOnly=pairedReadsOnly, minBaseQ=minBaseQuality, minMapQ=minMappingQuality)
-    val out     = makeWriter(output, vcfIn.header, filters)
+    val name = sample.getOrElse(VcfUtil.onlySample(source)).ensuring(
+      sample => source.header.samples.contains(sample),
+      s"There is no genotype with the following sample in the input VCF/BCF: $sample"
+    )
 
-    vcfIn.foreach { variant =>
-      val gt                = variant.genotypes(vcfSample)
-      val applicableFilters = filters.filter(_.appliesTo(gt))
+    val writer = VcfWriter(output, header = header)
 
-      val variantOut = if (applicableFilters.isEmpty) variant else {
-        val iterator = bamIn.query(variant.chrom, variant.pos, variant.pos, QueryType.Overlapping)
-        val pileup   = filterPileup(builder.build(iterator, variant.chrom, variant.pos).withoutOverlaps)
-        iterator.close()
+    source.tapEach(progress.record).foreach { variant =>
+      val updated = SomaticFilters.filter(_.appliesTo(variant.genotypes(name))) match {
+        case Nil     => variant
+        case filters =>
+          val pileup = if (streamBam) { piler.skipTo(refName = variant.chrom, pos = variant.pos) } else {
+            val overlapping = records.query(variant.chrom, start = variant.pos, end = variant.pos, Overlapping)
+            piler.build(overlapping, refName = variant.chrom, pos = variant.pos)
+          }
 
-        val allAnnotations = variant.attrs.toBuffer
-        val allFilters     = variant.filters.toBuffer
+          val nonOverlapping = pileup.withoutOverlaps
+          val allAnnotations = variant.attrs.toBuffer
+          val allFilters     = variant.filters.toBuffer
 
-        applicableFilters.foreach { f =>
-          // Generate the set of annotations, push them into the ctxBuilder, then also add any filters to the ctxBuilder
-          val annotations = f.annotations(pileup, gt)
-          allAnnotations ++= annotations
-          allFilters     ++= f.filters(annotations)
-        }
+          filters.foreach { filter =>
+            val annotations = filter.annotations(nonOverlapping, variant.genotypes(name))
+            allAnnotations ++= annotations
+            allFilters     ++= filter.filters(annotations)
+          }
 
-        variant.copy(attrs = ListMap(allAnnotations.toSeq:_*), filters = allFilters.toSet)
+          variant.copy(attrs = ListMap.from(allAnnotations), filters = Set.from(allFilters))
       }
-
-      out += variantOut
+      writer.write(updated)
     }
 
-    out.close()
-    vcfIn.safelyClose()
-    bamIn.safelyClose()
-  }
-
-  /**
-    * Filters the pileup to remove entries that are from PE reads who's insert size indicates that the
-    * aligned base we're looking at is _outside_ the insert!
-    */
-  private def filterPileup[A <: PileupEntry](pileup: Pileup[A]): Pileup[A] = {
```

Comment:
This feature was backwards implemented into `PileupBuilder` so it does not need to be here anymore.